Source/WebCore/html/canvas/WebGLProgram.cpp

```diff
@@ -111,16 +111,108 @@ void WebGLProgram::deleteObjectImpl(const AbstractLocker& locker, GraphicsContex
     }
 }
 
-bool WebGLProgram::getLinkStatus()
+bool WebGLProgram::linkStatus()
 {
-    cacheInfoIfNeeded();
-    return m_linkStatus;
+    if (!m_state.linkStatus) {
+        RefPtr context = graphicsContextGL();
+        if (!context)
+            return false;
+        m_state.linkStatus = context->getProgrami(object(), GraphicsContextGL::LINK_STATUS);
+    }
+    return *m_state.linkStatus;
+}
+
+std::span<const GCGLAttribActiveInfo> WebGLProgram::activeAttribs()
+{
+    if (!m_state.hasActiveAttribs) {
+        RefPtr context = graphicsContextGL();
+        if (!context)
+            return { };
+        m_state.activeAttribs = context->activeAttribs(object());
+        m_state.hasActiveAttribs = true;
+    }
+    return m_state.activeAttribs;
+}
+
+const HashMap<String, int>& WebGLProgram::attribLocations()
+{
+    if (!m_state.hasAttribLocations) {
+        auto& locations = m_state.attribLocations;
+        for (auto& activeAttrib : activeAttribs()) {
+            String name = String::fromUTF8(activeAttrib.name.span());
+            locations.add(name, activeAttrib.location);
+        }
+        m_state.hasAttribLocations = true;
+    }
+    return m_state.attribLocations;
+}
+
+std::span<const GCGLUniformActiveInfo> WebGLProgram::activeUniforms()
+{
+    if (!m_state.hasActiveUniforms) {
+        RefPtr context = graphicsContextGL();
+        if (!context)
+            return { };
+        m_state.activeUniforms = context->activeUniforms(object());
+        m_state.hasActiveUniforms = true;
+    }
+    return m_state.activeUniforms;
+}
+
+const HashMap<String, int>& WebGLProgram::uniformLocations()
+{
+    if (!m_state.hasUniformLocations) {
+        auto& locations = m_state.uniformLocations;
+        for (auto& activeUniform : activeUniforms()) {
+            auto name = String::fromUTF8(activeUniform.name.data());
+            locations.add(name, activeUniform.locations[0]);
+            if (name.endsWith("[0]"_s)) {
```

Comment:
This `[0]` suffix is removed in GraphicsContextGLANGLE::activeUniforms(). Should it be done only in one place; If we have to support non-Cocoa contexts, does it make sense to do it here?

Reply:
It is not removed in GraphicsContextGLANGLE. There it's removed to form the names to query the uniform locations of the subsequent array indices.

Here the condition is used to identify an array. E.g. "This active uniform is an array, so make it possible to query the array location with the base name in addition to fully qualified array first element name"